tools/swift-reflection-dump/swift-reflection-dump.cpp

```diff
@@ -32,200 +33,227 @@
 
 namespace options {
-static llvm::cl::opt<ActionType>
-Action(llvm::cl::desc("Mode:"),
-       llvm::cl::values(
-         clEnumValN(ActionType::DumpReflectionSections,
-                    "dump-reflection-sections",
-                    "Dump the field reflection section"),
-         clEnumValN(ActionType::DumpTypeLowering,
-                    "dump-type-lowering",
-                    "Dump the field layout for typeref strings read from stdin")),
-       llvm::cl::init(ActionType::DumpReflectionSections));
+static llvm::cl::opt<ActionType> Action(
+    llvm::cl::desc("Mode:"),
+    llvm::cl::values(
+        clEnumValN(ActionType::DumpReflectionSections,
+                   "dump-reflection-sections",
+                   "Dump the field reflection section"),
+        clEnumValN(
+            ActionType::DumpTypeLowering, "dump-type-lowering",
+            "Dump the field layout for typeref strings read from stdin")),
+    llvm::cl::init(ActionType::DumpReflectionSections));
 
 static llvm::cl::list<std::string>
-BinaryFilename("binary-filename", llvm::cl::desc("Filenames of the binary files"),
-               llvm::cl::OneOrMore);
+    BinaryFilename("binary-filename",
+                   llvm::cl::desc("Filenames of the binary files"),
+                   llvm::cl::OneOrMore);
 
 static llvm::cl::opt<std::string>
-Architecture("arch", llvm::cl::desc("Architecture to inspect in the binary"),
-             llvm::cl::Required);
+    Architecture("arch",
+                 llvm::cl::desc("Architecture to inspect in the binary"),
+                 llvm::cl::Required);
 } // end namespace options
 
-template<typename T>
-static T unwrap(llvm::Expected<T> value) {
+template <typename T> static T unwrap(llvm::Expected<T> value) {
   if (value)
     return std::move(value.get());
-  std::cerr << "swift-reflection-test error: " << toString(value.takeError()) << "\n";
+  llvm::errs() << "swift-reflection-test error: " << toString(value.takeError())
+               << "\n";
+  exit(EXIT_FAILURE);
+}
+
+static void reportError(std::error_code EC) {
+  assert(EC);
+  llvm::errs() << "swift-reflection-test error: " << EC.message() << ".\n";
   exit(EXIT_FAILURE);
 }
 
-using NativeReflectionContext
-    = swift::reflection::ReflectionContext<External<RuntimeTarget<sizeof(uintptr_t)>>>;
+using NativeReflectionContext =
+    ReflectionContext<External<RuntimeTarget<sizeof(uintptr_t)>>>;
+
+using ReadBytesResult = swift::remote::MemoryReader::ReadBytesResult;
+
+static uint64_t getSectionAddress(SectionRef S) {
+  if (auto C = dyn_cast<COFFObjectFile>(S.getObject()))
+    return S.getAddress() - C->getImageBase();
+  return S.getAddress();
+}
+
+class Image {
+  std::vector<char> Memory;
+
+public:
+  explicit Image(const ObjectFile *O) {
+    uint64_t Size = O->getData().size();
+    for (SectionRef S : O->sections()) {
+      if (auto SectionAddr = getSectionAddress(S))
+        Size = std::max(Size, SectionAddr + S.getSize());
```

Comment:
This is slightly confusing ... `Size` can be either "size" or the un-rebased end of the virtual address space.  Can we name this something to reflect that (`End` or `VASize`).  Additionally, this really could use a comment explaining that we need to compute the complete VA space that can be potentially occupied in order to properly compute the value of the relative pointers.